internal/jobcontainers/jobcontainer.go

```diff
@@ -166,10 +196,29 @@ func (c *JobContainer) CreateProcess(ctx context.Context, config interface{}) (_
 		return nil, errors.New("unsupported process config passed in")
 	}
 
-	// Replace any occurrences of the sandbox mount point env variable in the commandline.
-	// %CONTAINER_SANDBOX_MOUNTPOINT%\mybinary.exe -> C:\C\123456789\mybinary.exe
+	// Replace any occurences of the sandbox mount env variable in the commandline.
+	// For example: %CONTAINER_SANDBOX_MOUNTPOINT%\mybinary.exe -> C:\<rootfslocation>\mybinary.exe.
 	commandLine, _ := c.replaceWithMountPoint(conf.CommandLine)
 
+	// This is to workaround a rather unfortunate outcome with launching a process in a silo that
+	// has bound files.
+	// If a user requested to launch a program at C:\<rootfslocation>\mybinary.exe because they
+	// expect C:\<rootfslocation>\mybinary.exe to exist once the file bindings are done, it actually
+	// won't work. This is because the process is launched on the host first and the
+	// bindings are only visible to processes in that silo specifically. The shim is not
+	// one of these processes, and this is where we're invoking CreateProcess. So CreateProcess
+	// and our commandline resolution logic won't be able to find the process being asked for.
+	// Deep down in the depths of CreateProcess the culprit is a NtQueryAttributesFile call that
+	// fails as it doesn't have any context surrounding paths available to our silo.
+	//
+	// A way to get around this is to launch a process that will always exist (cmd) and is in our
+	// path, and then just invoke the program with the cmdline supplied. We could also add a new
+	// mode/flag for the shim where it's just a dummy process launcher, so we can invoke the shim
+	// instead of cmd and have more control over things.
+	if fileBindingSupport {
+		commandLine = "cmd /c " + commandLine
+	}
```

Comment:
@kevpar Curious if you think we should find the right approach with re-execing the shim as a dummy process launcher to handle this now. My idea was get this checked in first and we can leave the new shim functionality for last so it would all be in one change we can go over

Reply:
Using `cmd` as a proxy process seems acceptable for the time being, if we find problems later we can devise another approach to fix them.

I do think reinvoking the shim itself is probably better long-term. With cmd I'm especially concerned that it could impact us in weird ways, such as argument parsing.

Reply:
From talking with someone on the console team it didn't seem like it, although obviously there's now a new cmd/conhost overhead on each process which is crummy. The idea Kevin and I had floated around was adding new functionality to the shim so that it acts as a dummy process launcher. That way we could re-exec the shim inside the silo so it can actually see what we're trying to invoke, and then hopefully be able to exit the new dummy shim instance afterwards if we can redirect IO/wait on exit properly.

Reply:
will there be PTY/TTY issues, if the process detects its running in cmd and not enable certain features?

Reply:
> will there be PTY/TTY issues, if the process detects its running in cmd and not enable certain features?

I'd specifically asked about this (actually this was my first question 😆) and I didn't hear any concerns. 

> Using cmd as a proxy process seems acceptable for the time being, if we find problems later we can devise another approach to fix them.

> I do think reinvoking the shim itself is probably better long-term. With cmd I'm especially concerned that it could impact us in weird ways, such as argument parsing.

I'd like to swap to a shim re-exec approach before Containerd 1.7 hopefully. I'd feel better at least hahah